src/DataBuilder.php

```diff
@@ -743,38 +727,94 @@ protected function getBaseException()
         return $this->baseException;
     }
 
-    protected function getScrubFields()
+    public function getScrubFields()
     {
         return $this->scrubFields;
     }
+    
+    /**
+     * Scrub a data structure including arrays and query strings.
+     *
+     * @param mixed $data Data to be scrubbed.
+     * @param array $fields Sequence of field names to scrub.
+     * @param string $replacement Character used for scrubbing.
+     */
+    public function scrub(&$data, $replacement = '*')
+    {
+        $fields = $this->getScrubFields();
+        
+        if (!$fields || !$data) {
+            return $data;
+        }
+        
+        if (is_array($data)) { // scrub arrays
+        
+            $data = $this->scrubArray($data, $replacement);
+        } elseif (is_string($data)) { // scrub URLs and query strings
+            
+            $query = parse_url($data, PHP_URL_QUERY);
+            
+            /**
+             * String is not a URL but it still might be just a plain
+             * query string in format arg1=val1&arg2=val2
+             */
+            if (!$query) {
+                parse_str($data, $parsed);
+                $parsedValues = array_values($parsed);
+                
+                /**
+                 * If parsing a string results in an associative array
+                 * with multiple elements it's valid query string (key
+                 * recognition).
+                 *
+                 * Also, if it results in first key having an assigned value
+                 * it's also a valid query string (values recognition).
+                 */
+                if (count($parsed) > 1 || $parsedValues[0]) {
+                    $query = $data;
+                }
+            }
+                
+            if ($query) {
+                $data = str_replace(
+                    $query,
+                    $this->scrubQueryString($query),
+                    $data
+                );
+            }
+        }
+        
+        return $data;
+    }
 
-    protected function scrub($arr, $fields, $replacement = '*')
+    protected function scrubArray(&$arr, $replacement = '*')
     {
+        $fields = $this->getScrubFields();
+        
         if (!$fields || !$arr) {
-            return null;
+            return $arr;
         }
+        
+        $dataBuilder = $this;
 
-        $scrubber = function (&$val, $key) use ($fields, $replacement, $arr) {
-
-            if (key_exists($key, $arr)) {
+        $scrubber = function (&$val, $key) use ($fields, $replacement, &$scrubber, $dataBuilder) {
+            if (in_array($key, $fields, true)) {
                 $val = str_repeat($replacement, 8);
             }
+
+            $val = $dataBuilder->scrub($val, $replacement);
```

Comment:
Why not have this in an else block? It would seem that it is okay to not have it in an else because $val just ends up as the string '********' which scrub will just ignore, but why not avoid that call to scrub altogether?